ambry-account/src/main/java/com/github/ambry/account/mysql/DatasetDao.java

```diff
@@ -110,74 +111,80 @@ public DatasetDao(MySqlDataAccessor dataAccessor, MySqlAccountServiceConfig mySq
             DATASET_TABLE, ACCOUNT_ID, CONTAINER_ID, DATASET_NAME, VERSION_SCHEMA, LAST_MODIFIED_TIME, RETENTION_POLICY,
             RETENTION_COUNT, RETENTION_TIME_IN_SECONDS, USER_TAGS);
     updateDatasetIfExpiredSql = String.format(
-        "update %s set %s = ?, %s = now(3), %s = ?, %s = ?, %s = ?, %s = ?, %s = NULL where %s = ? and %s = ? and %s = ? and delete_ts < now(3)",
-        DATASET_TABLE, VERSION_SCHEMA, LAST_MODIFIED_TIME, RETENTION_POLICY, RETENTION_COUNT,
-        RETENTION_TIME_IN_SECONDS, USER_TAGS, DELETE_TS, ACCOUNT_ID, CONTAINER_ID, DATASET_NAME);
+        "update %1$s set %2$s = ?, %3$s = now(3), %4$s = ?, %5$s = ?, %6$s = ?, %7$s = ?, %8$s = NULL, %12$s = NULL "
```

Comment:
It's not about this PR. Just for this statement,
1. why do we set deleted_ts to NULL?
2. here we assume deleted_ts is not NULL. What if deleted_ts and delete_ts are both null?

Reply:
This statement is used when dataset is expired but has not been autopurged yet. So when user create a dataset with same datatset name, we should create a new dataset without setting any delete_ts or deleted_ts, since we don't support ttl for dataset(just for dataset version). The only way to delete the dataset is to call the deleteDataset API. You can see the addDataset will not update the delete_ts either. In summary, addDataset and updateDatasetIfExpired will make sure the delete_ts and deleted_ts in non-null.